### ISSUE TITLE
Loss Sanity Check

### ISSUE THREAD

**Issue by JoshVarty** [JoshVarty/LearningJax#2]
Something is broken. Let's figure out what.

- [x] After making all weights identical (#1), calculate loss on a batch of size 1.
- [x] Calculate loss on a batch of size 1 for Keras

Are they equal?

**Comment by JoshVarty** [comment]
The losses were similar but not equal. Further investigation showed that the Jax loss was incorrect due to an extra dimension on `logits` (shape `(5,1)`) that was not present on `rating` (shape `(5,)`). This lead to an incorrect loss and useless gradients.

We fixed this by squeezing `x` after concatenating the movie and user embeddings in the forward pass.

**Comment by JoshVarty** [comment]
Re-opening because we should honestly be able to get bitwise equivalent results for the loss (assuming we use the logits from the keras model). We need to understand why this is not the case.

- There is a cast (and `batch_ratings` is 64-bit)

Where is tf.math.squared_difference implemented?
